src/Raven.Server/Indexing/VoronIndexInput.cs

```diff
@@ -10,6 +11,221 @@
+        private bool _isDisposed = false;
+
+        private bool _isOriginal = true;
+
+        public VoronBufferedInput(LuceneVoronDirectory directory, string name, Transaction transaction, string tree) : base(DefaultBufferSize)
+        {
+            _directory = directory;
+            _name = name;
+            _tree = tree;
+
+            OpenInternal(transaction);
+            
+            //We don't want to get buffer instantly. Only when needed.
+            //buffer = ArrayPool<byte>.Shared.Rent(DefaultBufferSize);
+        }
+
+        public override string ToString()
+        {
+            return _name;
+        }
+
+        public override void SetBufferSize(int newSize)
+        {
+            if (buffer == null)
+                return;
+            
+            if (newSize <= buffer.Length)
+                return;
+
+            var newBuffer = ArrayPool<byte>.Shared.Rent(newSize);
+            Array.Copy(buffer, newBuffer, buffer.Length);
+            var oldBuffer = buffer;
+            buffer = newBuffer;
+            ArrayPool<byte>.Shared.Return(oldBuffer);
+        }
+
+        private void OpenInternal(Transaction transaction)
+        {
+            if (transaction.IsWriteTransaction == false)
+            {
+                if (transaction.LowLevelTransaction.ImmutableExternalState is IndexTransactionCache cache)
+                {
+                    if (cache.DirectoriesByName.TryGetValue(_directory.Name, out var files))
+                    {
+                        if (files.ChunksByName.TryGetValue(_name, out var details))
+                        {
+                            // we don't dispose here explicitly, the fileName needs to be
+                            // alive as long as the transaction is
+                            Slice.From(transaction.Allocator, _name, out Slice fileName);
+                            _stream = new VoronStream(fileName, details, transaction.LowLevelTransaction);
+                            return;
+                        }
+                    }
+                }
+            }
+
+            var fileTree = transaction.ReadTree(_tree);
+            if (fileTree == null)
+                throw new FileNotFoundException($"Could not find '{_tree}' tree for index input", _name);
+
+            using (Slice.From(transaction.Allocator, _name, out Slice fileName))
+            {
+                _stream = fileTree.ReadStream(fileName);
+                if (_stream == null)
+                    throw new FileNotFoundException("Could not find index input", _name);
+            }
+        }
+
+        public override object Clone(IState s)
+        {
+            var state = s as VoronState;
+            if (state == null)
+            {
+                ThrowStateNullException();
+                return null;
+            }
+
+            ThrowIfDisposed(state);
+
+            var clone = (VoronBufferedInput)base.Clone(s);
+            GC.SuppressFinalize(clone);
```

Comment:
An opened question here is whatever we want to clone the buffer as well. I'm leaning toward yet.

Reply:
https://github.com/maciejaszyk/ravendb/blob/dd0038910c991bb74de84af25dd6c29e876d1605/templucene/Lucene.Net/Store/BufferedIndexInput.cs#L230-L240

Lucene is deleting reference by default.